File: backend/app/schemas/orchestrator.py
import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field
from app.models.model_registry import ModelType


class OrchestratorPredictRequest(BaseModel):
    transaction_id: str = Field(..., example="TXN-8D93240D", description="Target Transaction ID")
    requested_model_types: Optional[List[ModelType]] = Field(
        None, description="Optional subset of model types to orchestrate. If omitted, runs all active models."
    )


class CompositePredictionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    prediction_id: str
    transaction_id: str
    feature_vector_id: Optional[str] = None
    overall_risk_score: float
    confidence: float
    composite_risk_level: str
    executed_models: List[str]
    execution_time_ms: float
    individual_results: Dict[str, Any]
    feature_version: str
    model_versions: Dict[str, str]
    metadata_json: Dict[str, Any]
    created_at: datetime
    updated_at: datetime


class PaginatedCompositePredictionResponse(BaseModel):
    items: List[CompositePredictionResponse]
    total: int
    page: int
    size: int
    pages: int

File: frontend/src/types/orchestrator.ts
export interface IndividualModelResult {
  model_id: string;
  model_name: string;
  framework: string;
  raw_result: string; // ALLOW, FLAG, BLOCK
  score: number; // 0.0 to 100.0
  confidence: number;
  latency_ms: number;
  status: string; // SUCCESS, FAILED, SKIPPED, FALLBACK
  error_message?: string;
}

export interface CompositePredictionRecord {
  id: string;
  prediction_id: string;
  transaction_id: string;
  feature_vector_id?: string;
  overall_risk_score: number;
  confidence: number;
  composite_risk_level: string; // LOW, MEDIUM, HIGH, CRITICAL
  executed_models: string[];
  execution_time_ms: number;
  individual_results: Record<string, IndividualModelResult>;
  feature_version: string;
  model_versions: Record<string, string>;
  metadata_json: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface PaginatedCompositePredictions {
  items: CompositePredictionRecord[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface OrchestratorFilterParams {
  transaction_id?: string;
  composite_risk_level?: string;
  page?: number;
  size?: number;
  sort_by?: string;
  sort_dir?: string;
}
